python/shotgun_model/shotgun_query_model.py

```diff
@@ -0,0 +1,947 @@
+
+        SG_DATA_UNIQUE_ID_FIELD = "id"
+
+    """
+
+    # ---- signals
+
+    # signal emitted after the model's sg query is changed
+    query_changed = QtCore.Signal()
+
+    # signal emitted after the model loads cache data
+    cache_loaded = QtCore.Signal()
+
+    # signal emitted before the model starts to refresh its shotgun data
+    data_refreshing = QtCore.Signal()
+
+    # signal emitted after the model is updated with fresh shotgun data
+    data_refreshed = QtCore.Signal(bool)
+
+    # signal emitted in the case the refresh fails
+    data_refresh_fail = QtCore.Signal(str)
+
+    # ---- internal constants
+
+    # please do not access directly but instead use the helper
+    # methods provided! We may change these constants without prior notice.
+    SG_DATA_ROLE = QtCore.Qt.UserRole + 1
+    IS_SG_MODEL_ROLE = QtCore.Qt.UserRole + 2
+
+    # ---- "abstract" class members
+
+    # subclasses should define a unique id field from the SG data for use in
+    # caching and associating data with items in the model
+    SG_DATA_UNIQUE_ID_FIELD = None
+
+    # subclasses should define the class to use when loading items from disk
+    SG_QUERY_MODEL_ITEM_CLASS = ShotgunStandardItem
+
+    # ---- caching/serialization related costants
+
+    # magic number for IO streams
+    _FILE_MAGIC_NUMBER = 0xDEADBEEF
+
+    # version of binary format
+    _FILE_VERSION = 22
+
+    def __init__(self, parent, bg_task_manager=None):
+        """
+        Initializes the model and provides some default convenience members.
+
+        :param parent: The model's parent.
+        :type parent: :class:`~PySide.QtGui.QObject`
+
+        :param bg_task_manager: Background task manager to use for any
+            asynchronous work. If this is None then a task manager will be
+            created as needed.
+        :type bg_task_manager: :class:`~task_manager.BackgroundTaskManager`
+
+        The following instance members are created for use in subclasses:
+
+        :protected _bundle: The current toolkit bundle
+
+        :protected _shotgun_data: ``shotgunutils.shotgun_data`` handle
+
+        :protected _shotgun_globals: ``shotgunutils.shotgun_globals`` handle
+
+        :protected _sg_data_retriever: A ``ShotgunDataRetriever`` instance.
+            Connected to virtual slots ``_on_data_retriever_work_completed``
+            and ``_on_data_retriever_work_failure`` which subclasses must
+            implement.
+
+        """
+
+        # ensure subclasses define the required class members
+        if not self.SG_DATA_UNIQUE_ID_FIELD:
+            raise ShotgunModelError(
+                "ShotgunQueryModel subclass does not define the instance attr: "
+                "`SG_DATA_UNIQUE_ID_FIELD`"
+            )
+
+        # intialize the Qt base class
+        super(ShotgunQueryModel, self).__init__(parent)
+
+        # keep a handle to the current app/engine/fw bundle for convenience
+        self._bundle = sgtk.platform.current_bundle()
```

Comment:
given the subclassing going on here, i think we should be very explicit about encapuslation - e.g. all private member variables having the double underscore - single underscore denoting that we can access that variable in deriving classes. Does that make sense?

So the default should be `self.__double_underscore` and for members that we want to expose to deriving classes, we do `self._single_underscore` (but possibly better to do it via properties?)